drv/ignition-server/src/main.rs

```diff
@@ -0,0 +1,243 @@
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+        link: LinkSelect,
+    ) -> Result<LinkEvents, RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        self.controller
+            .link_events(port, link)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+
+    fn clear_link_events(
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+        link: LinkSelect,
+    ) -> Result<(), RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        self.controller
+            .clear_link_events(port, link)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+
+    fn send_request(
+        &mut self,
+        _: &userlib::RecvMessage,
+        port: u8,
+        request: Request,
+    ) -> Result<(), RequestError> {
+        if port >= self.port_count {
+            return Err(RequestError::from(IgnitionError::InvalidPort));
+        }
+
+        let port_state = self.port_state(port).map_err(RequestError::from)?;
+        let target_state = port_state
+            .target()
+            .ok_or(RequestError::from(IgnitionError::NoTargetPresent))?;
+
+        if target_state.system_reset_in_progress()
+            || target_state.system_power_off_in_progress()
+            || target_state.system_power_on_in_progress()
+        {
+            return Err(RequestError::from(IgnitionError::RequestInProgress));
+        }
+
+        self.controller
+            .set_request(port, request)
+            .map_err(IgnitionError::from)
+            .map_err(RequestError::from)
+    }
+}
+
+impl idol_runtime::NotificationHandler for ServerImpl {
+    fn current_notification_mask(&self) -> u32 {
+        TIMER_NOTIFICATION_MASK
+    }
+
+    fn handle_notification(&mut self, _bits: u32) {
+        let start = sys_get_timer().now;
+
+        // Only poll the presence summary if the port count seems reasonable. A
+        // count of 0xff may occur if the FPGA is running an incorrect
+        // bitstream.
+        if self.port_count > 0 && self.port_count != 0xff {
+            if let Err(_e) = self.poll_presence() {}
+        }
+
+        let finish = sys_get_timer().now;
+
+        // We now know when we were notified and when any work was completed.
+        // Note that the assumption here is that `start` < `finish` and that
+        // this won't hold if the system time rolls over. But, the system timer
+        // is a u64, with each bit representing a ms, so in practice this should
+        // be fine. Anyway, armed with this information, find the next deadline
+        // some multiple of `TIMER_INTERVAL` in the future.
+
+        let delta = finish - start;
+        let next_deadline = finish + TIMER_INTERVAL - (delta % TIMER_INTERVAL);
```

Comment:
This is a little funky – what's the advantage versus doing `let next_deadline = finish + TIMER_INTERVAL`?

Reply:
`let next_deadline = finish + TIMER_INTERVAL` guarantees `TIMER_INTERVAL` between notifications while the method above stays close to a specific periodic point in time, irrespective of how long work takes, at the expense of skipping a "frame" if work pushes past what would have been the next deadline.

This code comes from the sequencer where I feel this matters more. What's here will be replaced by interrupts soon, so it's probably fine either way.